### PR TITLE
fix: Multitenancy schema updates

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change
(A few sentences about this PR)

## Related issues
- Link to issue1 here
- Link to issue1 here

## Test Plan
(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes
(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
   - In `build.gradle`
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        });
+    }
+
+    public static void addTenantIdInUserPool(Start start, TenantIdentifier tenantIdentifier) throws
+            StorageTransactionLogicException, StorageQueryException {
+        {
+            start.startTransaction(con -> {
+                Connection sqlCon = (Connection) con.getConnection();
+                try {
+                    {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:141]: use the same `System.currentTimeMillis()` for both the queries below.

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 2]
```diff
@@ -2033,13 +2033,33 @@ public HashMap<String, String> getUserIdMappingForSuperTokensIds(AppIdentifier a
...
+            MultitenancyQueries.createTenant(this, tenantConfig);
+        } catch (SQLException e) {
+            PostgreSQLConfig config = Config.getConfig(this);
+            if (isPrimaryKeyError(((PSQLException) e).getServerErrorMessage(), config.getTenantConfigsTable())) {
+                throw new DuplicateTenantException();
+            }
+            // TODO: what about other error?
+        } catch (StorageQueryException e) {
+            // TODO: what about other error?
+        }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2047]:
```suggestion
       
```

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 3]
```diff
@@ -2033,13 +2033,33 @@ public HashMap<String, String> getUserIdMappingForSuperTokensIds(AppIdentifier a
...
-        // TODO:
+    public void createTenant(TenantConfig tenantConfig) throws DuplicateTenantException {
+        try {
+            MultitenancyQueries.createTenant(this, tenantConfig);
+        } catch (SQLException e) {
+            PostgreSQLConfig config = Config.getConfig(this);
+            if (isPrimaryKeyError(((PSQLException) e).getServerErrorMessage(), config.getTenantConfigsTable())) {
+                throw new DuplicateTenantException();
+            }
+            // TODO: what about other error?
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2044]:
```suggestion
            throw new StorageQueryException(e);
```

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 4]
```diff
@@ -2033,13 +2033,33 @@ public HashMap<String, String> getUserIdMappingForSuperTokensIds(AppIdentifier a
     }
 
     @Override
-    public void createTenant(TenantConfig config) throws DuplicateTenantException {
-        // TODO:
+    public void createTenant(TenantConfig tenantConfig) throws DuplicateTenantException {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2036]:
```suggestion
    public void createTenant(TenantConfig tenantConfig) throws DuplicateTenantException, StorageQueryException {
```

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 5]
```diff
@@ -2049,7 +2069,13 @@ public void deleteTenantIdInUserPool(TenantIdentifier tenantIdentifier) throws T
 
     @Override
     public void overwriteTenantConfig(TenantConfig config) throws TenantOrAppNotFoundException {
-        // TODO:
+        try {
+            MultitenancyQueries.updateTenant(this, config);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2073]:
```suggestion
            MultitenancyQueries. overwriteTenantConfig(this, config);
```

Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2073]: you need to check if number of rows affected are 0, then you need to throw TenantOrAppNotFoundException

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 6]
```diff
@@ -2033,13 +2033,33 @@ public HashMap<String, String> getUserIdMappingForSuperTokensIds(AppIdentifier a
     }
 
     @Override
-    public void createTenant(TenantConfig config) throws DuplicateTenantException {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2036]: you should test all these storage functions with all the expected types of errors thrown as well (in the core)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 7]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+
+                    update(sqlCon, QUERY2, pst -> {
+                        pst.setString(1, tenantConfig.tenantIdentifier.getConnectionUriDomain());
+                        pst.setString(2, tenantConfig.tenantIdentifier.getAppId());
+                        pst.setString(3, tenantConfig.tenantIdentifier.getTenantId());
+                        pst.setString(4, provider.thirdPartyId);
+                        pst.setString(5, providerClient.clientType);
+                        pst.setString(6, providerClient.clientId);
+                        pst.setString(7, providerClient.clientSecret);
+                        pst.setArray(8, sqlCon.createArrayOf("CHAR", providerClient.scope));
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:317]:
```suggestion
                        pst.setArray(8, sqlCon.createArrayOf("VARCHAR", providerClient.scope));
```

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 8]
```diff
@@ -2033,32 +2033,28 @@ public HashMap<String, String> getUserIdMappingForSuperTokensIds(AppIdentifier a
...
-            MultitenancyQueries.createTenant(this, tenantConfig);
-        } catch (SQLException e) {
+            MultitenancyQueries.createTenantConfig(this, tenantConfig);
+        } catch (StorageTransactionLogicException e) {
             PostgreSQLConfig config = Config.getConfig(this);
-            if (isPrimaryKeyError(((PSQLException) e).getServerErrorMessage(), config.getTenantConfigsTable())) {
+            if (isPrimaryKeyError(((PSQLException) e.actualException).getServerErrorMessage(), config.getTenantConfigsTable())) {
                 throw new DuplicateTenantException();
             }
-            // TODO: what about other error?
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java]: you need to add more exception checks here

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 9]
```diff
@@ -2068,13 +2064,14 @@ public void deleteTenantIdInUserPool(TenantIdentifier tenantIdentifier) throws T
     }
 
     @Override
-    public void overwriteTenantConfig(TenantConfig config) throws TenantOrAppNotFoundException {
+    public void overwriteTenantConfig(TenantConfig config) throws TenantOrAppNotFoundException, StorageQueryException {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2067]: test this function by creating a thread pool of 300 threads and then all of them in parallel querying this function -> check that all succeed.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/postgresql/Start.java:2067]: See this test for reference: https://github.com/supertokens/supertokens-postgresql-plugin/blob/master/src/test/java/io/supertokens/storage/postgresql/test/DeadlockTest.java#L157

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 10]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
-            pst.setString(7, tenantConfig.thirdPartyConfig.toJSON().toString());
+    static String getQueryToCreateTenantThirdPartyProvidersTable(Start start) {
+        String schema = Config.getConfig(start).getTableSchema();
+        String tenantThirdPartyProvidersTable = Config.getConfig(start).getTenantThirdPartyProvidersTable();
+        // @formatter:off
+        return "CREATE TABLE IF NOT EXISTS " + tenantThirdPartyProvidersTable + " ("
+                + "connection_uri_domain CHAR(256) DEFAULT '',"
+                + "app_id CHAR(64) DEFAULT 'public',"
+                + "tenant_id CHAR(64) DEFAULT 'public',"
+                + "third_party_id CHAR(64) NOT NULL,"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:176]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 11]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
-            pst.setString(3, tenantConfig.tenantIdentifier.getTenantId());
-            pst.setString(4, tenantConfig.coreConfig.toString());
-            pst.setString(5, tenantConfig.emailPasswordConfig.toJSON().toString());
-            pst.setString(6, tenantConfig.passwordlessConfig.toJSON().toString());
-            pst.setString(7, tenantConfig.thirdPartyConfig.toJSON().toString());
+    static String getQueryToCreateTenantThirdPartyProvidersTable(Start start) {
+        String schema = Config.getConfig(start).getTableSchema();
+        String tenantThirdPartyProvidersTable = Config.getConfig(start).getTenantThirdPartyProvidersTable();
+        // @formatter:off
+        return "CREATE TABLE IF NOT EXISTS " + tenantThirdPartyProvidersTable + " ("
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:172]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 12]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + "connection_uri_domain CHAR(256) DEFAULT '',"
+                + "app_id CHAR(64) DEFAULT 'public',"
+                + "tenant_id CHAR(64) DEFAULT 'public',"
+                + "third_party_id CHAR(64) NOT NULL,"
+                + "name CHAR(64),"
+                + "authorization_endpoint CHAR(256),"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:178]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 13]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + "authorization_endpoint_query_params TEXT,"
+                + "token_endpoint CHAR(256),"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:180]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 14]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + "token_endpoint_body_params TEXT,"
+                + "user_info_endpoint CHAR(256),"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:182]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 15]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + "user_info_endpoint_query_params TEXT,"
+                + "user_info_endpoint_headers TEXT,"
+                + "jwks_uri CHAR(256),"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:185]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 16]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + "oidc_discovery_endpoint CHAR(256),"
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:186]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 17]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                + " FOREIGN KEY(connection_uri_domain, app_id, tenant_id)"
+                + " REFERENCES " + Config.getConfig(start).getTenantConfigsTable() +  " (connection_uri_domain, app_id, tenant_id) ON DELETE CASCADE"
+                + ");";
+        // @formatter:on
+    }
+
+    static String getQueryToCreateTenantThirdPartyProviderClientsTable(Start start) {
+        String schema = Config.getConfig(start).getTableSchema();
+        String tenantThirdPartyProvidersTable = Config.getConfig(start).getTenantThirdPartyProviderClientsTable();
+        return "CREATE TABLE IF NOT EXISTS " + tenantThirdPartyProvidersTable + " ("
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:205]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 18]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                    {
+                        String QUERY = "DELETE FROM " + getConfig(start).getTenantThirdPartyProvidersTable()
+                                + " WHERE connection_uri_domain = ? AND app_id = ? AND tenant_id = ?;";
+                        update(start, QUERY, pst -> {
+                            pst.setString(1, tenantConfig.tenantIdentifier.getConnectionUriDomain());
+                            pst.setString(2, tenantConfig.tenantIdentifier.getAppId());
+                            pst.setString(3, tenantConfig.tenantIdentifier.getTenantId());
+                        });
+                    }
+                    */
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:368]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 19]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                    {
+                        String QUERY = "DELETE FROM " + getConfig(start).getTenantConfigsTable()
+                                + " WHERE connection_uri_domain = ? AND app_id = ? AND tenant_id = ?;";
+                        int rowsAffected = update(start, QUERY, pst -> {
+                            pst.setString(1, tenantConfig.tenantIdentifier.getConnectionUriDomain());
+                            pst.setString(2, tenantConfig.tenantIdentifier.getAppId());
+                            pst.setString(3, tenantConfig.tenantIdentifier.getTenantId());
+                        });
+                        if (rowsAffected == 0) {
+                            sqlCon.rollback();
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:379]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 20]
```diff
@@ -69,81 +69,461 @@ public TenantConfig map(ResultSet result) throws StorageQueryException {
...
+                pst.setBoolean(5, tenantConfig.emailPasswordConfig.enabled);
+                pst.setBoolean(6, tenantConfig.passwordlessConfig.enabled);
+                pst.setBoolean(7, tenantConfig.thirdPartyConfig.enabled);
+            });
+        }
+
+        {
+            if (tenantConfig.thirdPartyConfig.providers == null) {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:243]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java (hunk 7 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:317]: Done

File: src/main/java/io/supertokens/storage/postgresql/Start.java (hunk 9 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/Start.java:2067]: Done

File: src/main/java/io/supertokens/storage/postgresql/Start.java (hunk 8 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/Start.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/Start.java [hunk 21]
```diff
@@ -2064,13 +2074,22 @@ public void deleteTenantIdInUserPool(TenantIdentifier tenantIdentifier) throws T
...
+            if (e.actualException instanceof TenantOrAppNotFoundException) {
                 throw (TenantOrAppNotFoundException) e.actualException;
             }
+            PostgreSQLConfig config = Config.getConfig(this);
+            if (isPrimaryKeyError(((PSQLException) e.actualException).getServerErrorMessage(), config.getTenantThirdPartyProvidersTable())) {
+                throw new DuplicateThirdPartyIdException();
+            }
+            if (isPrimaryKeyError(((PSQLException) e.actualException).getServerErrorMessage(), config.getTenantThirdPartyProviderClientsTable())) {
+                throw new DuplicateClientTypeException();
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/Start.java:2092]:
```suggestion
```

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 22]
```diff
@@ -238,86 +258,82 @@ private static void executeCreateTenantQueries(Start start, Connection sqlCon, T
...
+
+                            update(sqlCon, QUERY2, pst -> {
+                                pst.setString(1, tenantConfig.tenantIdentifier.getConnectionUriDomain());
+                                pst.setString(2, tenantConfig.tenantIdentifier.getAppId());
+                                pst.setString(3, tenantConfig.tenantIdentifier.getTenantId());
+                                pst.setString(4, provider.thirdPartyId);
+                                pst.setString(5, providerClient.clientType);
+                                pst.setString(6, providerClient.clientId);
+                                pst.setString(7, providerClient.clientSecret);
+                                pst.setArray(8, sqlCon.createArrayOf("VARCHAR", providerClient.scope));
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:331]: seems wrong

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java (hunk 22 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java]: It was fine, however, I have updated to make the code consistent.

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Pull request merged**